flink-table/flink-table-planner-blink/src/main/java/org/apache/calcite/sql/fun/SqlDotOperator.java

```diff
@@ -0,0 +1,194 @@
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+import org.apache.calcite.sql.validate.SqlValidatorUtil;
+import org.apache.calcite.util.Litmus;
+import org.apache.calcite.util.Static;
+
+import java.util.Arrays;
+
+/**
+ * The dot operator {@code .}, used to access a field of a
+ * record. For example, {@code a.b}.
+ *
+ * <p>This class was copied over from Calcite to fix the derived type.
+ * If the ROW is nullable force the accessed field to be nullable as well.
+ */
+public class SqlDotOperator extends SqlSpecialOperator {
+  SqlDotOperator() {
+    super("DOT", SqlKind.DOT, 100, true, null, null, null);
+  }
+
+  @Override public ReduceResult reduceExpr(int ordinal, TokenSequence list) {
+    SqlNode left = list.node(ordinal - 1);
+    SqlNode right = list.node(ordinal + 1);
+    return new ReduceResult(ordinal - 1,
+        ordinal + 2,
+        createCall(
+            SqlParserPos.sum(
+                Arrays.asList(left.getParserPosition(),
+                    right.getParserPosition(),
+                    list.pos(ordinal))),
+            left,
+            right));
+  }
+
+  @Override public void unparse(SqlWriter writer, SqlCall call, int leftPrec,
+      int rightPrec) {
+    final SqlWriter.Frame frame =
+        writer.startList(SqlWriter.FrameTypeEnum.IDENTIFIER);
+    call.operand(0).unparse(writer, leftPrec, 0);
+    writer.sep(".");
+    call.operand(1).unparse(writer, 0, 0);
+    writer.endList(frame);
+  }
+
+  @Override public SqlOperandCountRange getOperandCountRange() {
+    return SqlOperandCountRanges.of(2);
+  }
+
+  @Override public <R> void acceptCall(SqlVisitor<R> visitor, SqlCall call,
+      boolean onlyExpressions, SqlBasicVisitor.ArgHandler<R> argHandler) {
+    if (onlyExpressions) {
+      // Do not visit operands[1] -- it is not an expression.
+      argHandler.visitChild(visitor, call, 0, call.operand(0));
+    } else {
+      super.acceptCall(visitor, call, onlyExpressions, argHandler);
+    }
+  }
+
+  @Override public RelDataType deriveType(SqlValidator validator,
+      SqlValidatorScope scope, SqlCall call) {
+    final SqlNode operand = call.getOperandList().get(0);
+    final RelDataType nodeType =
+        validator.deriveType(scope, operand);
+    assert nodeType != null;
+    if (!nodeType.isStruct()) {
+      throw SqlUtil.newContextException(operand.getParserPosition(),
+          Static.RESOURCE.incompatibleTypes());
+    }
+
+    final SqlNode fieldId = call.operand(1);
+    final String fieldName = fieldId.toString();
+    final RelDataTypeField field =
+        nodeType.getField(fieldName, false, false);
+    if (field == null) {
+      throw SqlUtil.newContextException(fieldId.getParserPosition(),
+          Static.RESOURCE.unknownField(fieldName));
+    }
+    RelDataType type = field.getType();
+    if (nodeType.isNullable() && !type.isNullable()) {
+        type = validator.getTypeFactory().createTypeWithNullability(type, true);
+    }
```

Comment:
In Calcite, when the record type is nullable, the attributes are always nullable, so, this fix is not that necessary from the Calcite side.